File: Pommora/Pommora/Vaults/PropertyValue.swift
import Foundation

/// Type-erased property value used in Item / Page / Agenda `properties` dictionaries.
/// Custom Codable inspects the JSON shape per-value:
/// - JSON number → `.number(Double)`
/// - JSON bool   → `.checkbox(Bool)`
/// - JSON null   → `.null`
/// - JSON array of objects `[{"$rel": "..."}, ...]` → `.relation([String])` (ULIDs of target entities)
/// - JSON object `{"$status": "..."}` → `.status(String)` (option value)
/// - JSON array of strings → `.multiSelect([String])`
/// - JSON array of objects → `.file([FileRef])`
/// - JSON string → `.url`/`.date`/`.datetime`/`.select` (disambiguated by shape;
///                  ISO-8601 strings decode as `.datetime` if they include time, `.date` if not;
///                  URLs validate via `URL(string:)`; anything else is `.select`)
///
/// Relation encoding: relations are always multi-valued. `.relation(ids)` writes an array of
/// tagged objects `[{"$rel": id}, ...]` (a single target is a one-element array) so external
/// agents and the graph-view indexer can identify cross-entity edges from any single file
/// without consulting the Type schema. A legacy single `{"$rel": id}` object still decodes to
/// `.relation([id])`. Satisfies Pommora load-bearing constraint #3.
///
/// Status encoding: `.status(value)` writes `{"$status": value}` for the same reason — it
/// disambiguates status (grouped picker) from `.select` (flat picker) at the value layer,
/// since both store a single option string. Properties.md describes the conceptual on-disk
/// shape (`"<option value>"`); the manager layer can translate between schema-aware string
/// form and tagged form, but the Codable layer needs the tag for round-trip stability.
///
/// Date vs datetime: `.date` writes a yyyy-MM-dd string (UTC); `.datetime` writes full
/// ISO-8601 with timezone. On decode: ISO-8601 with `T` → `.datetime`, else yyyy-MM-dd → `.date`.
///
/// `.lastEditedTime` is a **virtual** case — it is never stored on disk (the value derives from
/// the file's `modified_at` at read time). Encoding `.lastEditedTime` throws `EncodingError.invalidValue`
/// to prevent accidental persistence.
enum PropertyValue: Codable, Equatable, Hashable, Sendable {
    case number(Double)
    case checkbox(Bool)
    case date(Date)
    case datetime(Date)
    case select(String)
    case multiSelect([String])
    case status(String)  // option value; encodes as {"$status": value}
    case relation([String])  // ULIDs of target entities; encodes as [{"$rel": id}, ...]; tier-only tolerance; retired from user creation
    case url(URL)
    case file([FileRef])
    case lastEditedTime  // virtual — never persisted; encoding throws
    case null

    // MARK: - Codable

    init(from decoder: any Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
            return
        }
        if let b = try? c.decode(Bool.self) {
            self = .checkbox(b)
            return
        }
        if let n = try? c.decode(Double.self) {
            self = .number(n)
            return
        }
        // Relation array (always-multi): `[{"$rel": id}, ...]`. Must precede the
        // [FileRef] checks — FileRef object-arrays lack a `$rel` key so they won't
        // match here, and a bare `[String]` isn't `[[String: String]]`. The non-empty
        // guard excludes `[]` so empty arrays keep falling through to `.file([])`.
        if let arr = try? c.decode([[String: String]].self),
            !arr.isEmpty,
            arr.allSatisfy({ $0["$rel"] != nil })
        {
            self = .relation(arr.compactMap { $0["$rel"] })
            return
        }
        // Array branch: prefer file (array of objects) over multi-select (array of strings)
        if let files = try? c.decode([FileRef].self), !files.isEmpty {
            self = .file(files)
            return
        }
        if let arr = try? c.decode([String].self) {
            self = .multiSelect(arr)
            return
        }
        // Empty array → treat as empty file list (multi-select empty arrays should not occur
        // at the storage layer; if needed, the manager can normalise).
        if let empties = try? c.decode([FileRef].self), empties.isEmpty {
            self = .file([])
            return
        }
        // Tagged-object: {"$rel": "01H..."} or {"$status": "value"}
        // Legacy single `$rel` object decodes to a one-element relation array.
        if let obj = try? c.decode([String: String].self), obj.count == 1 {
            if let id = obj["$rel"] {
                self = .relation([id])
                return
            }
            if let value = obj["$status"] {
                self = .status(value)
                return
            }
        }
        if let s = try? c.decode(String.self) {
            // Try URL
            if let url = URL(string: s), url.scheme != nil {
                self = .url(url)
                return
            }
            // Try ISO-8601 datetime
            let isoDateTime = ISO8601DateFormatter()
            isoDateTime.formatOptions = [.withInternetDateTime]
            if let d = isoDateTime.date(from: s) {
                self = .datetime(d)
                return
            }
            // Try yyyy-MM-dd
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd"
            dateFormatter.timeZone = TimeZone(identifier: "UTC")
            if let d = dateFormatter.date(from: s) {
                self = .date(d)
                return
            }
            // Fallthrough: plain string → treat as select value
            self = .select(s)
            return
        }
        throw DecodingError.dataCorruptedError(
            in: c,
            debugDescription: "PropertyValue: unrecognised JSON shape"
        )
    }

    func encode(to encoder: any Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .number(let n): try c.encode(n)
        case .checkbox(let b): try c.encode(b)
        case .select(let s): try c.encode(s)
        case .multiSelect(let xs): try c.encode(xs)
        case .status(let value): try c.encode(["$status": value])
        case .relation(let ids): try c.encode(ids.map { ["$rel": $0] })
        case .url(let u): try c.encode(u.absoluteString)
        case .file(let refs): try c.encode(refs)
        case .null: try c.encodeNil()
        case .date(let d):
            let f = DateFormatter()
            f.dateFormat = "yyyy-MM-dd"
            f.timeZone = TimeZone(identifier: "UTC")
            try c.encode(f.string(from: d))
        case .datetime(let d):
            let iso = ISO8601DateFormatter()
            iso.formatOptions = [.withInternetDateTime]
            try c.encode(iso.string(from: d))
        case .lastEditedTime:
            throw EncodingError.invalidValue(
                self,
                EncodingError.Context(
                    codingPath: encoder.codingPath,
                    debugDescription:
                        "PropertyValue.lastEditedTime is virtual and must not be persisted; "
                        + "derive from the file's modified_at at read time."
                )
            )
        }
    }
}

File: React/src/shared/propertyValue.ts
// PropertyValue — the type-erased on-disk value used in Page/Agenda `properties`.
// The DECLARED type lives in the schema; this codec recovers a value from raw JSON
// by SHAPE, in a LOCKED precedence that mirrors the Swift original exactly. The
// order is load-bearing and silent on failure — reordering a branch mistypes
// contexts/files/multi-select with no error. Pure: no fs, no Node — importable by
// both main and renderer.

/** On-disk file-attachment shape (snake_case = the on-disk DTO). Round-trips as-is;
 *  unknown keys on a file object are preserved (the codec passes the object through). */
export interface FileRef {
  path: string
  original_name?: string
  added_at?: string
  mime_type?: string
}

export type PropertyValue =
  | { kind: 'number'; value: number }
  | { kind: 'checkbox'; value: boolean }
  | { kind: 'datetime'; value: string } // ISO-8601; a bare "yyyy-MM-dd" is a date-only datetime
  | { kind: 'select'; value: string }
  | { kind: 'multiSelect'; value: string[] }
  | { kind: 'status'; value: string }
  | { kind: 'context'; value: string[] } // context-tier target ULIDs
  | { kind: 'url'; value: string }
  | { kind: 'file'; value: FileRef[] }
  | { kind: 'lastEditedTime' } // virtual — never persisted (encode throws)
  | { kind: 'null' }

// RFC-3986 scheme prefix — matches Swift's `URL(string:).scheme != nil`.
const SCHEME = /^[a-zA-Z][a-zA-Z0-9+.-]*:/
const YMD = /^\d{4}-\d{2}-\d{2}$/
// ISO-8601 with time + timezone designator — matches Swift's `.withInternetDateTime`
// (no fractional seconds; a TZ is required, else the string falls through to select).
const ISO_DATETIME = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:Z|[+-]\d{2}:\d{2})$/

/** True for a plain (non-null, non-array) object. The one shared shape guard for JSON /
 *  frontmatter records across the data layer. */
export function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

function isFileRef(v: unknown): v is FileRef {
  return isPlainObject(v) && typeof v.path === 'string'
}

/**
 * Classify a raw JSON value into a PropertyValue. BRANCH ORDER IS LOAD-BEARING —
 * mirrors Swift `PropertyValue.init(from:)`:
 * null → bool → number → non-empty [{$ctx}] → non-empty [FileRef] → [string]
 * (incl. empty [] → multiSelect([])) → single {$ctx}/{$status} →
 * string(url → iso-datetime → yyyy-MM-dd-as-datetime → select). Anything else throws.
 */
export function parsePropertyValue(raw: unknown): PropertyValue {
  if (raw === null || raw === undefined) return { kind: 'null' }
  if (typeof raw === 'boolean') return { kind: 'checkbox', value: raw }
  if (typeof raw === 'number') return { kind: 'number', value: raw }

  if (Array.isArray(raw)) {
    if (raw.length > 0 && raw.every((x) => isPlainObject(x) && typeof x.$ctx === 'string')) {
      return { kind: 'context', value: raw.map((x) => (x as { $ctx: string }).$ctx) }
    }
    if (raw.length > 0 && raw.every(isFileRef)) {
      return { kind: 'file', value: raw }
    }
    if (raw.every((x) => typeof x === 'string')) {
      // Includes the empty array: `[]` → multiSelect([]) (Swift's [String] branch
      // catches it before the dead file([]) path).
      return { kind: 'multiSelect', value: raw as string[] }
    }
    throw new Error('PropertyValue: unrecognised array shape')
  }

  if (isPlainObject(raw)) {
    const keys = Object.keys(raw)
    if (keys.length === 1) {
      if (typeof raw.$ctx === 'string') return { kind: 'context', value: [raw.$ctx] }
      if (typeof raw.$status === 'string') return { kind: 'status', value: raw.$status }
    }
  }

  if (typeof raw === 'string') {
    if (SCHEME.test(raw)) return { kind: 'url', value: raw }
    if (ISO_DATETIME.test(raw)) return { kind: 'datetime', value: raw }
    if (YMD.test(raw)) return { kind: 'datetime', value: raw } // a bare date is a date-only datetime
    return { kind: 'select', value: raw }
  }

  throw new Error('PropertyValue: unrecognised JSON shape')
}

/** Encode a PropertyValue to its on-disk JSON value. The switch is exhaustive (the
 *  compiler enforces every case). `lastEditedTime` is virtual and throws. */
export function encodePropertyValue(value: PropertyValue): unknown {
  switch (value.kind) {
    case 'number':
      return value.value
    case 'checkbox':
      return value.value
    case 'select':
      return value.value
    case 'url':
      return value.value
    case 'datetime':
      return value.value
    case 'multiSelect':
      return value.value
    case 'status':
      return { $status: value.value }
    case 'context':
      return value.value.map((id) => ({ $ctx: id }))
    case 'file':
      return value.value
    case 'null':
      return null
    case 'lastEditedTime':
      throw new Error(
        'PropertyValue.lastEditedTime is virtual and must not be persisted; derive from modified_at at read time.'
      )
  }
}

/** True when a value carries nothing — an empty array or empty string. Checkbox `false` and
 *  number `0` are real values and stay. */
function isEmptyValue(value: PropertyValue): boolean {
  switch (value.kind) {
    case 'multiSelect':
    case 'context':
    case 'file':
      return value.value.length === 0
    case 'select':
    case 'status':
    case 'url':
    case 'datetime':
      return value.value === ''
    default:
      return false
  }
}

/** Set or clear one property on a (possibly malformed) properties record, returning the next
 *  record. A null value (the `null` kind) OR an empty value clears the key — a page without a
 *  value has no key at all, never a null/[]/'' placeholder — anything else encodes via the
 *  codec. The single owner of the page + agenda property set/clear rule. */
export function applyPropertyValue(
  current: unknown,
  propertyId: string,
  value: PropertyValue | null
): Record<string, unknown> {
  const next: Record<string, unknown> = isPlainObject(current) ? { ...current } : {}
  if (value === null || value.kind === 'null' || isEmptyValue(value)) delete next[propertyId]
  else next[propertyId] = encodePropertyValue(value)
  return next
}
